### PR TITLE
Feature/199

### LINKED ISSUE #192: Update server applications
Now  volumemanager,  workflows and common are running on flask server.
Use the schema provided by harness-application script (see also "samples" application) to run with gunicorn

### PR DESCRIPTION
**Pull request by filippomc**
This PR includes:

-  integration with Skaffold, providing better dev experience, with faster/targeted builds, auto deployment
- refactoring of applications docker files to comply with skaffold requirements
- debug capability with cloud code/VS code
- refactoring of python flask applications (and generation) with new centralized cloud-harness initialization

# In-cluster debug with VS code

Remote debug can be achieved through Skaffold by using the Cloud Code extension for Visual studio code.
All required configurations for Skaffold and VS code are generated by the harness-deployment command.

1. Setup kubectl to use your remote cluster or minikube*. Set the default namespace to be the same namespace you are making the deployment (the same specified on running harness-deployment)
1. Install Skaffold: see https://skaffold.dev/docs/install/
1. Install Cloud Code extension on VS Code: https://cloud.google.com/code/docs/vscode/quickstart-local-dev#installing
1. Launch the "CloudHarness: run/debug configuration"
image

# Build/deploy with skaffold

[Skaffold](https://skaffold.dev/) configuration is generated by `harness-deployment`.

Can build and deploy with Skaffold with:

`skaffold dev --namespace=[NAMESPACE]` (local minikube)

`skaffold dev --namespace=osb2 --default-repo=[DOCKER_REGISTRY]` (external cluster or forwarded minikube)

Closes #192 #199 #17

### CODE REVIEW

File: applications/volumemanager/server/Dockerfile [hunk 1]
```diff
@@ -1,21 +1,20 @@
...
 COPY requirements.txt /usr/src/app/
 
 RUN pip3 install --no-cache-dir -r requirements.txt
 
 COPY . /usr/src/app
 
-EXPOSE 8080
-
-ENTRYPOINT ["python3"]
+RUN pip3 install -e .
```
Comment (zsinnema) [inline applications/volumemanager/server/Dockerfile:18]: why install the working folder in edit mode?

File: applications/volumemanager/server/volumemanager/__main__.py [hunk 2]
```diff
@@ -1,18 +1,9 @@
...
-
-def main():
-    app = connexion.App(__name__, specification_dir='./openapi/')
-    app.app.json_encoder = encoder.JSONEncoder
-    app.add_api('openapi.yaml',
-                arguments={'title': 'Volumes manager API'},
-                pythonic_params=True)
-    app.run(port=8080)
 
+init_flask()
```
Comment (zsinnema) [inline applications/volumemanager/server/volumemanager/__main__.py:6]: there is no "app" in __main__.py but it is used in the Dockerfile entrypoint

File: applications/workflows/server/Dockerfile [hunk 3]
```diff
@@ -1,18 +1,18 @@
...
+ENV MODULE_NAME=workflows_api
+
+COPY ./requirements.txt /usr/src/app/
 
 RUN pip3 install --no-cache-dir -r requirements.txt
 
 COPY . /usr/src/app
 
 EXPOSE 8080
+RUN pip3 install -e .
```
Comment (zsinnema) [inline applications/workflows/server/Dockerfile:16]: why in "edit" mode?

File: applications/workflows/server/workflows_api/__main__.py [hunk 4]
```diff
@@ -1,18 +1,10 @@
...
-def main():
-    app = connexion.App(__name__, specification_dir='./openapi/')
-    app.app.json_encoder = encoder.JSONEncoder
-    app.add_api('openapi.yaml',
-                arguments={'title': 'Workflows API'},
-                pythonic_params=True)
-    app.run(port=8080)
+from cloudharness.utils.server import init_flask, main
 
+init_flask()
```
Comment (zsinnema) [inline applications/workflows/server/workflows_api/__main__.py:5]: also no app var in__main__ see dockerfile entrypoint

File: infrastructure/base-images/cloudharness-base-debian/Dockerfile [hunk 5]
```diff
@@ -1,11 +1,15 @@
...
+COPY libraries/client/cloudharness_cli/requirements.txt /libraries/client/cloudharness_cli/requirements.txt
+RUN pip install -r /libraries/client/cloudharness_cli/requirements.txt
+
 COPY libraries/cloudharness-common /libraries/cloudharness-common
 COPY libraries/client/cloudharness_cli /libraries/client/cloudharness_cli
 
-RUN pip install -r /libraries/cloudharness-common/requirements.txt
-RUN pip install /libraries/cloudharness-common
-RUN pip install /libraries/client/cloudharness_cli
+RUN pip install -e /libraries/cloudharness-common
```
Comment (zsinnema) [inline infrastructure/base-images/cloudharness-base-debian/Dockerfile:12]: why in edit mode?

File: infrastructure/base-images/cloudharness-base-debian/Dockerfile [hunk 6]
```diff
@@ -1,11 +1,15 @@
...
+RUN pip install -e /libraries/client/cloudharness_cli
```
Comment (zsinnema) [inline infrastructure/base-images/cloudharness-base-debian/Dockerfile:13]: same

File: infrastructure/base-images/cloudharness-base/Dockerfile [hunk 7]
```diff
@@ -8,10 +8,15 @@ RUN apk upgrade
...
+COPY libraries/cloudharness-common/requirements.txt /libraries/cloudharness-common/requirements.txt
+RUN pip install -r /libraries/cloudharness-common/requirements.txt
+COPY libraries/client/cloudharness_cli/requirements.txt /libraries/client/cloudharness_cli/requirements.txt
+RUN pip install -r /libraries/client/cloudharness_cli/requirements.txt
 
 COPY libraries/cloudharness-common /libraries/cloudharness-common
 COPY libraries/client/cloudharness_cli /libraries/client/cloudharness_cli
 
+#
 RUN pip install /libraries/cloudharness-common
```
Comment (zsinnema) [inline infrastructure/base-images/cloudharness-base/Dockerfile:20]: and here not in edit mode

File: utilities/cloudharness_utilities/application-templates/server/backend/__APP_NAME__/__main__.py [hunk 8]
```diff
@@ -1,37 +1,10 @@
 #!/usr/bin/env python3
 
-import logging
+from cloudharness.utils.server import init_flask, main
 
-import connexion
-import flask
-from __APP_NAME__ import encoder
 
+init_flask(title="__APP_NAME__", init_app_fn=None, webapp=False)
```
Comment (zsinnema) [inline utilities/cloudharness_utilities/application-templates/server/backend/__APP_NAME__/__main__.py:6]: here I would do 
```
app = init_flask(....)
```
so we can use app in the Dockerfile

File: utilities/cloudharness_utilities/application-templates/webapp/backend/__APP_NAME__/__main__.py [hunk 9]
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env python3
+
+from cloudharness.utils.server import init_flask, main
+
+
+init_flask(title="__APP_NAME__", init_app_fn=None, webapp=True)
```
Comment (zsinnema) [inline utilities/cloudharness_utilities/application-templates/webapp/backend/__APP_NAME__/__main__.py:6]:
```
app = init_flask(...)
```

**Review by zsinnema — CHANGES_REQUESTED** [review]

**Comment by zsinnema** [comment]
@filippomc also the skaffold.yaml deploy part is missing the namespace of the release
```
    releases:
    - name: xxx
      chartPath: deployment/helm
      **namespace: <--- this is missing**
      skipBuildDependencies: true
      artifactOverrides:
```

File: applications/common/server/Dockerfile [hunk 10]
```diff
@@ -1,24 +1,22 @@
...
-RUN apk update && apk add postgresql-dev gcc python3-dev musl-dev
-#RUN apk add psycopg2 libpq-dev python-dev
+FROM $CLOUDHARNESS_FLASK
 
-RUN mkdir -p /usr/src/app
-WORKDIR /usr/src/app
+ENV MODULE_NAME=common
 
-COPY requirements.txt /usr/src/app/
+ENV WORKERS=2
```
[the remark this answers is not in the document]
Comment (filippomc) [reply applications/common/server/Dockerfile:9]: https://docs.gunicorn.org/en/0.16.1/design.html#how-many-workers there are reasons to have 2 at least

File: applications/volumemanager/server/Dockerfile (hunk 1 above)
Comment (filippomc) [reply applications/volumemanager/server/Dockerfile:18]: It's simpler to track the debugger files mapping if we don't move the files

File: applications/common/server/requirements.txt [hunk 11]
```diff
@@ -9,7 +9,7 @@ swagger-ui-bundle >= 0.0.2
 python_dateutil >= 2.6.0
 setuptools >= 21.0.0
 
-psycopg2-binary==2.8.5
+psycopg2
```
[the remark this answers is not in the document]
Comment (filippomc) [reply applications/common/server/requirements.txt:12]: psycopg2-binary does not found

**Review by zsinnema — APPROVED** [review]
lgtm

**Pull request merged**